staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -112,34 +112,37 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
 }
 
 type listenAddress struct {
-	address     string
-	protocol    string
-	failureMode string
+	address  string
+	protocol string
 }
 
+// parseAddresses parses, validates addresses
+// Keeps the order from addressesToParse, but removes duplicates
 func parseAddresses(addressesToParse []string) ([]listenAddress, error) {
 	var addresses []listenAddress
-	parsed := make(map[string]listenAddress)
+	parsed := make(map[string]bool)
+
 	for _, address := range addressesToParse {
-		if address == "localhost" {
-			ip := listenAddress{address: "127.0.0.1", protocol: "tcp4", failureMode: "all"}
-			parsed[ip.address] = ip
-			ip = listenAddress{address: "::1", protocol: "tcp6", failureMode: "all"}
-			parsed[ip.address] = ip
-		} else if net.ParseIP(address).To4() != nil {
-			parsed[address] = listenAddress{address: address, protocol: "tcp4", failureMode: "any"}
-		} else if net.ParseIP(address) != nil {
-			parsed[address] = listenAddress{address: address, protocol: "tcp6", failureMode: "any"}
-		} else {
-			return nil, fmt.Errorf("%s is not a valid IP", address)
+		ips, err := net.LookupIP(address)
```

Comment:
unfortunately LookupIP is not a reliable way of getting loopback address(es). It relies on name resolution, which in turn will use ie. /etc/hosts so even if you have both `127.0.0.1` and `::1` but only one of them in hosts file, this method will report only the IP defined in hosts.

https://github.com/kubernetes/kubernetes/pull/69578/files#diff-43e7f6a6df5c4f7000f26222d9afb0cfR399 is the solution suggested by @lavalamp for this.

Reply:
Thanks for looking into this.

I have some concerns about this approach in context of `kubectl`, but l I'm not an expert in this, so, please, excuse me if I say something stupid :)

In the code example above, the function gets the loopback interface by hardcoded name `lo`. How will this work on different operating systems? On Linux machines I see `lo` interface, but on my Mac OS I see `lo0` interface. Will it work on Windows?

I think we can loop over all interfaces and check if an interface is has both `FlagUp` and `FlagLoopback` flags to avoid hardcoding the name, but this raises more questions:

1. What if we find more than one loopback interface?
2. Should we try to bind to all IPs from all interfaces we find?
3. If users explicitly runs `--address localhost`, do resolve using DNS or do we look for IPs on loopback interfaces?

Reply:
the implementation in original code is for `localhost` to simply translate to `127.0.0.1` and/or `::1`. I don't think we should support any other addresses for non-explicit, default `localhost`. If you want to bind to some IP on unusual loopback, you can just provide the IP in the `--address` flag. Maybe just search if each of these two localhost addresses exist on some `FlagUp` interface.

Btw. LookupIP has another side effect of reintroducing ability to use names in `--address` values, while it was explicitly removed in original PR as was seen as expanding the scope of kubectl / client responsibility too much (like using `--address some.domain.tld`).

Wonder if it wouldn't be the best to return an error like `can't bind twice to the same address` if the IP shows up twice (ie. from `127.0.0.1,localhost`) instead of updating and messing up failure mode in this edge case.

Reply:
> Maybe just search if each of these two localhost addresses exist on some `FlagUp` interface.

Makes perfect sense. I can do that.

> Wonder if it wouldn't be the best to return an error like `can't bind twice to the same address` if the IP shows up twice (ie. from `127.0.0.1,localhost`) instead of updating and messing up failure mode in this edge case.

I like the idea of returning the "can't bind twice" error if for some reason the same address appears twice. Let's be explicit.

---

Before I go further, do I understand correctly that we still want to allow users something like `--address localhost,10.19.21.23` (it's from `kubectl port-forward --help` message)? There were suggestions in https://github.com/kubernetes/kubernetes/pull/46517#discussion_r223808638 and https://github.com/kubernetes/kubernetes/pull/46517#discussion_r179140223 from @lavalamp and @liggitt about not allowing passing `localhost` into `--address` (if I interpreted their comments correctly).

Dropping this will make the implementation much clearer, but I'm afraid that v1.13.0-rc.1 has already been released and it's too late to drop it.

Reply:
> the implementation in original code is for localhost to simply translate to 127.0.0.1 and/or ::1. I don't think we should support any other addresses for implicit, default localhost. If you want to bind to some IP on unusual loopback, you can just provide the IP in the --address flag.

I agree with that, and the behavior should be documented in the flag help.

>> Maybe just search if each of these two localhost addresses exist on some FlagUp interface.
>
> Makes perfect sense. I can do that.

I'd actually try to reduce the amount of interface inspection done, not increase it. I think the following behavior would be reasonable:

1. explicitly specified addresses must bind successfully
2. `localhost` is pinned to `127.0.0.1` and `::1` and at least one of those must bind successfully

2 can be accomplished by changing the localhost address parsing to never override an explicitly specified address:

  ```go
  if address == "localhost" {
    if _, exists := parsed["127.0.0.1"]; !exists {
      ip := listenAddress{address: "127.0.0.1", protocol: "tcp4", failureMode: "all"}
      parsed[ip.address] = ip
    }
    if _, exists := parsed["::1"]; !exists {
      ip = listenAddress{address: "::1", protocol: "tcp6", failureMode: "all"}
      parsed[ip.address] = ip
    }
  } ...
  ```